pkg/machines/libvirt-dbus.es6

```diff
@@ -0,0 +1,861 @@
+                    });
+        };
+    },
+
+    CREATE_AND_ATTACH_VOLUME({
+        connectionName,
+        poolName,
+        volumeName,
+        size,
+        format,
+        target,
+        vmId,
+        permanent,
+        hotplug
+    }) {
+        let volXmlDesc = getVolumeXML(volumeName, size, format, target);
+
+        return dispatch => {
+            clientLibvirt[connectionName].call('/org/libvirt/QEMU', 'org.libvirt.Connect', 'StoragePoolLookupByName', [poolName], TIMEOUT)
+                    .done((storagePoolPath) => {
+                        clientLibvirt[connectionName].call(storagePoolPath[0], 'org.libvirt.StoragePool', 'StorageVolCreateXML', [volXmlDesc, 0], TIMEOUT)
+                                .done((storageVolumePath) => {
+                                    clientLibvirt[connectionName].call(storageVolumePath[0], "org.freedesktop.DBus.Properties", "GetAll", ["org.libvirt.StorageVol"], TIMEOUT)
+                                            .done(function(props) {
+                                                if (!$.isEmptyObject(props[0]))
+                                                    return dispatch(attachDisk({ connectionName, diskFileName: props[0].Path.v.v, target, vmId, permanent, hotplug }))
+                                                            .then(() => {
+                                                                // force reload of VM data, events are not reliable (i.e. for a down VM)
+                                                                dispatch(getVm({connectionName, id:vmId}));
+                                                            });
+                                            })
+                                            .fail(function(ex) { console.warn("failed waiting fot the StorageVol proxy to get ready", ex) });
+                                })
+                                .fail(ex => console.error("StorageVolCreateXML for pool %s failed: %s", poolName, ex));
+                    })
+                    .fail(ex => console.error("StoragePoolLookupByName for pool %s failed:", poolName, ex));
+        };
+    },
+
+    DELETE_VM({
+        name,
+        connectionName,
+        id: objPath,
+        options
+    }) {
+        function destroy() {
+            return clientLibvirt[connectionName].call(objPath, 'org.libvirt.Domain', 'Destroy', [0], TIMEOUT);
+        }
+
+        function undefine() {
+            (function undefine_synced(i) {
+                if (i < options.storage.length)
+                    clientLibvirt[connectionName].call('/org/libvirt/QEMU', 'org.libvirt.Connect', 'StorageVolLookupByPath', [options.storage[i]], TIMEOUT)
+                            .done((storageVolPath) => {
+                                clientLibvirt[connectionName].call(storageVolPath[0], 'org.libvirt.StorageVol', 'Delete', [0], TIMEOUT)
+                                        .finally(undefine_synced.bind(null, i + 1));
```

Comment:
This looks a bit curious. Deleting the storage volumes in parallel isn't safe? Or do they depend on each other somehow? If so, is it guaranteed that the dependencies are in numerical order? This could use a comment.

Reply:
We can delete storage volumes in parallel. I just need to make sure that after deleting the storage all storage volumes I 'll Undefine the VM. You suggest putting all promises for volume deletion in some array and calling  Promise.al? Or what is the proper way to do it?

Reply:
Right, that's the usual approach. It also (usually) makes error handling easier, as `Promise.all()` has a sensible behaviour (succeeds if everything succeeds, or fails if anything fails)

Reply:
ok thanks and sorry for my ignorance, javascript total newbie here.

Reply:
FTR, this can be written more elegantly using `promises = mylist.map()`, but this isn't a blocker.